trie/utils.go

```diff
@@ -29,24 +29,38 @@ package trie
 // This tool can track all of them no matter the node is embedded in its
 // parent or not, but valueNode is never tracked.
 //
+// Besides, it's also used for recording the original value of the nodes
+// when they are resolved from the disk. The pre-value of the nodes will
+// be used to construct reverse-diffs in the future.
+//
 // Note tracer is not thread-safe, callers should be responsible for handling
 // the concurrency issues by themselves.
 type tracer struct {
 	insert map[string]struct{}
 	delete map[string]struct{}
+	origin map[string][]byte
 }
 
-// newTracer initializes trie node diff tracer.
+// newTracer initializes the tracer for capturing trie changes.
 func newTracer() *tracer {
 	return &tracer{
 		insert: make(map[string]struct{}),
 		delete: make(map[string]struct{}),
+		origin: make(map[string][]byte),
+	}
+}
+
+// onRead tracks the newly loaded trie node and caches the rlp-encoded blob internally.
```

Comment:
This does not copy the value. Might be worth mentioning.

Reply:
Done